balancer-js/src/modules/data/pool/subgraphOnChain.ts

```diff
@@ -0,0 +1,112 @@
+import { Findable, Searchable } from '../types';
+import { Provider } from '@ethersproject/providers';
+import { PoolAttribute, PoolsRepositoryFetchOptions } from './types';
+import { GraphQLQuery, Pool } from '@/types';
+import { Network } from '@/lib/constants/network';
+import { getOnChainBalances } from '../../../modules/sor/pool-data/onChainData';
+import { PoolsSubgraphRepository } from './subgraph';
+
+interface PoolsSubgraphOnChainRepositoryOptions {
+  url: string;
+  chainId: Network;
+  provider: Provider;
+  multicall: string;
+  vault: string;
+  blockHeight?: () => Promise<number | undefined>;
+  query?: GraphQLQuery;
+}
+
+/**
+ * Access pools using generated subgraph client and multicall.
+ */
+export class PoolsSubgraphOnChainRepository
+  implements Findable<Pool, PoolAttribute>, Searchable<Pool>
+{
+  private poolsSubgraph: PoolsSubgraphRepository;
+  private provider: Provider;
+  private pools?: Promise<Pool[]>;
+  private multicall: string;
+  private vault: string;
+  public skip = 0;
+
+  /**
+   * Repository using multicall to get onchain data.
+   *
+   * @param url subgraph URL
+   * @param chainId current network, needed for L2s logic
+   * @param blockHeight lazy loading blockHeigh resolver
+   * @param multicall multicall address
+   * @param valt vault address
+   */
+  constructor(options: PoolsSubgraphOnChainRepositoryOptions) {
+    this.poolsSubgraph = new PoolsSubgraphRepository({
```

Comment:
I was struggling a bit to find a way to avoid repeating code here. Jarek suggested creating an `OnChainProvider` that would do something pretty similar to what's being done here. It's nice that I'll be able to use that once it's merged 😃